fixes issue #420 Add an optional string id to listener and filter

KSchopmeyer opened the pull request:
Ready for review:

Andreas: you may elect to put this on off until you finish restructuring listener

Adds an optional user defined string id to the listener
definition an another one to the filter definition.

The listener id is prepended to the name of all destination and
listener objects if it exists.  The filter id is prepended to the
name property of all filters after the listener id.

These are an informal mechanism for the user to identify instances
of filters and destination ojects that it has registered with the
WBEM server.

Nothing is prepended if these attributes are None

andy-maier commented on pywbem/_listener.py:
```diff
@@ -632,6 +632,11 @@ def __init__(self, host, http_port=DEFAULT_LISTENER_PORT_HTTP,
+
+          listener_id (:term:`string`):
+            If not `None` a string that prefixes the name property of filters
+            and handler instances to help the listener identify its filters and
```
On the term "handler": We use the term "listener destination" elsewhere, and I suggest to stay consistent. If you think it is needed, we could introduce the term "handler" somewhere, e.g. in `WBEMListener.add_server()`.

Related to consistency: I realized that in `add_server()`, the description uses the term "indication listener", which is also inconsistent and should be changed to "listener destination".

andy-maier commented on pywbem/_listener.py:
```diff
@@ -632,6 +632,11 @@ def __init__(self, host, http_port=DEFAULT_LISTENER_PORT_HTTP,
+          listener_id (:term:`string`):
+            If not `None` a string that prefixes the name property of filters
+            and handler instances to help the listener identify its filters and
+            handler instances in a WBEMServer.
```
On term "WBEMServer": Here we talk about the actual WBEM server, not about the class `WBEMServer`, so I suggest to change this to "WBEM server.".

andy-maier commented on pywbem/_listener.py:
```diff
@@ -893,6 +905,9 @@ def add_server(self, server):
                              server.conn.url)
+                             
+        # set the prefix into the server object
+        server.listener_id = self._listener_id
```
I don't like that at all; it creates a new instance variable from outside of the class, that has no real semantic meaning for that class, and that uses the object just as a store to pass it on to when it is needed.

Alternative proposal, removing the need for adding the `listener_id` instance variable to `WBEMServer`:
* For `_create_destination()`, add a `listener_id` parameter, and when calling it in `add_server()`, pass `self.__listener_id` as its value.
* For `_create_filter()`, add a `listener_id` parameter, and when calling it in `add_dynamic_filter()`, pass `self.__listener_id` as its value.

andy-maier commented on pywbem/_listener.py:
```diff
@@ -1250,13 +1265,15 @@ def _create_destination(server, dest_url):
 
-def _create_filter(server, source_namespace, query, query_language):
+def _create_filter(server, source_namespace, query, query_language, \
+                   indication_id_prefix=None):
```
The `indication_id_prefix` parameter is not used in `add_dynamic_filter()` which is the only place that calls this internal function. As a result, the user so far has no way to define it.

If it is really necessary to have this ID in addition to the listener ID, it should be externalized in `add_dynamic_filter()`. If not, it should be removed.

andy-maier commented:
-1
See comments.

KSchopmeyer commented:
Incorporated comments above and extended the pegasus tests to test most of the id alternatives.

KSchopmeyer commented:
+1

andy-maier commented:
+1
Except, you could add a test for the name strings.
Please rebase.

KSchopmeyer commented:
Added more tests so that now the IDs formatted names are completely tested.  Also cleaned up other pegasus based listener tests and one issue in the listener itself.  This passes the pywbem tests and also the pegasus tests.

+1
